### PR TITLE
grpc access logging: initial implementation

### PR DESCRIPTION
**Pull request by mattklein123**
This is an initial implementation for gRPC access logging. There are
quite a few TODOs sprinkled throughout this change that I or someone
else at Lyft will follow up on but this is good enough to kick the
tires and do some initial testing.

*Risk Level*: Low (new feature)
*Testing*: Unit/integration
*Docs Changes*: Will do in follow up
*Release Notes*: Aded
Part of https://github.com/envoyproxy/envoy/issues/743

### CODE REVIEW

**Comment by mattklein123** [comment]
@goaway @junr03 @lita @rodaine PTAL

@lita @rodaine you will likely be doing a bunch of the TODOs so please look at those also.

**Comment by mattklein123** [comment]
@dnoe @mrice32 if you could also take a brief look that would be cool. Not sure if any of this will be useful at Google or not.

**Comment by mrice32** [comment]
cc @tschroed

**Comment by mattklein123** [comment]
@ccaraman can you please review this next week. Thank you.

File: source/server/config/access_log/grpc_access_log.cc [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+#include "api/filter/accesslog/accesslog.pb.validate.h"
+
+namespace Envoy {
+namespace Server {
+namespace Configuration {
+
+class GrpcAccessLogClientFactoryImpl : public AccessLog::GrpcAccessLogClientFactory {
+public:
+  GrpcAccessLogClientFactoryImpl(Upstream::ClusterManager& cluster_manager,
```
Comment (ramaraochavali) [inline source/server/config/access_log/grpc_access_log.cc]: Where is the access log server cluster referred by this  cluster_name configured? i could not find such config in boostrap config. Is it done differently here?

**Comment by lita** [comment]
In the future, we might want to add a filter to log GRPC errors separately as well.

**Comment by mattklein123** [comment]
> In the future, we might want to add a filter to log GRPC errors separately as well.

+1 agreed

File: test/integration/access_log_integration_test.cc [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  AccessLogIntegrationTest() : HttpIntegrationTest(Http::CodecClient::Type::HTTP1, GetParam()) {}
+
+  void createUpstreams() override {
+    HttpIntegrationTest::createUpstreams();
+    fake_upstreams_.emplace_back(new FakeUpstream(0, FakeHttpConnection::Type::HTTP2, version_));
+  }
+
+  void initialize() override {
+    config_helper_.addConfigModifier([](envoy::api::v2::Bootstrap& bootstrap) {
+      bootstrap.mutable_rate_limit_service()->set_cluster_name("accesslog");
```
Comment (ccaraman) [inline test/integration/access_log_integration_test.cc]: curious: Why is the ratelimit service being used here?

File: test/common/access_log/grpc_access_log_impl_test.cc [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+          port_value: 0
+      start_time:
+        seconds: 3600
+      time_to_last_rx_byte:
+        nanos: 2000000
+      time_to_first_upstream_rx_byte:
+        nanos: 4000000
+      time_to_last_downstream_tx_byte:
+        nanos: 6000000
+      time_to_last_downstream_tx_byte:
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply test/common/access_log/grpc_access_log_impl_test.cc]: bug

File: test/integration/access_log_integration_test.cc (hunk 2 above)
Comment (mattklein123) [reply test/integration/access_log_integration_test.cc]: bug/copy+paste

**Comment by mattklein123** [comment]
@dnoe @mrice32 @tschroed do any of you want to review this prior to merge?

File: source/server/config/access_log/grpc_access_log.cc (hunk 1 above)
Comment (ramaraochavali) [reply source/server/config/access_log/grpc_access_log.cc]: ok. But how does envoy know that it should connect to that cluster? is it by some standard name? If you LDS, Ratelimiting etc, they have specific configuration in bootstrap.yaml that tells what cluster should be used for them. or Am I missing some thing here?

Comment (mattklein123) [reply source/server/config/access_log/grpc_access_log.cc]: It's not a filter, but an access log provider which has a config. Take a look at the integration test. That has a complete config.

**Pull request merged**